### ISSUE TITLE
RFC: Check and fail on non-required dependencies

### ISSUE THREAD

**Issue by adalessandro** [mesonbuild/meson#7013]
AFAICS, in order to fail if a non-required dependency is missing, `dependency.found()` has to be called and then raise the error by calling `assert()` or `error()`. For instance:

**Source files**
`meson.build`:
```meson
project('test', 'c')

foo_dep = dependency('foo_dep', required : false)

# NOTE: using a subdir and an option to emphasize that `foo_dep`
# should be non-required and the dependency logic should be kept
# inside the subdir target.
if get_option('foo_option') # Boolean option
    subdir('foo_exec')
endif
```

`foo_exec/meson.build`:
```meson
# Case failing with assert()
assert(foo_dep.found(), 'Dependency "foo_dep" not found')

# Case failing with error()
if not foo_dep.found()
    error('Dependency "foo_dep" not found')
endif

foo_exec = executable('foo_exec',
                      'foo_lib.c',
                      dependencies : foo_dep)
```

**Build output**
```bash
$ python --version
Python 3.8.2

$ meson build/ 
The Meson build system
Version: 0.54.999
[...]
Build type: native build
Project name: test
Project version: undefined
C compiler for the host machine: ccache cc (gcc 9.3.0 "cc (Arch Linux 9.3.0-1) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (1.6.3)
Found CMake: /usr/bin/cmake (3.16.5)
Run-time dependency foo_dep found: NO (tried pkgconfig and cmake)

foo_exec/meson.build:1:0: ERROR: Assert failed: Dependency "foo_dep" not found
[...]
```

**Possible enhancement**
Currently, the error message has to be written manually. It would be a good idea to add a new method to the `dependency` object, so it fails with the default error message (as if it were required) if not found.

For the lack of a better name, let's just call the method `.assert()`:

**Code example and build output**

`foo_exec/meson.build`:
```meson
# Print the same error message as calling:
#     dependency('foo_dep', required : true)
foo_dep.assert()

foo_exec = executable('foo_exec',
                      'foo_lib.c',
                      dependencies : foo_dep)
```

```bash
$ meson build/ 
[...]
Run-time dependency foo_dep found: NO (tried pkgconfig and cmake)

foo_exec/meson.build:1:0: ERROR: Dependency "foo_dep" not found, tried pkgconfig and cmake
[...]
```

I'd like to have some comments on this approach. Eventually, I could put a MR together to implement it.

**Comment by eli-schwartz** [comment]
https://mesonbuild.com/Build-options.html#features

```
foo_dep = dependency('foo_dep', required : get_option('foo_option'))

if get_option('foo_option').enabled() # Feature option
    subdir('foo_exec')
endif
```

Alternative implementation:
```
if get_option('foo_option') # Boolean option
    foo_dep = dependency('foo_dep')
    subdir('foo_exec')
endif
```

Third option: put the dependency() in the subdir.

**Comment by adalessandro** [comment]
Hi @eli-schwartz , thanks for the quick reply. Comments below:

> https://mesonbuild.com/Build-options.html#features

I understand about how features work, however I don't think that fully solves the problem described.

> ```
> foo_dep = dependency('foo_dep', required : get_option('foo_option'))
> 
> if get_option('foo_option').enabled() # Feature option
>     subdir('foo_exec')
> endif
> ```

This is not suitable if you have multiple subdirs that can depend on `foo_dep` and other dependencies, as you'd end with very long `if` clauses describing all the dependency logic for each case in the top `meson.build` script. Also, in my example, the dependency option is intentionally decoupled from the subdir option, i.e. could be enabled independently and should fail if enabled but dependencies are not satisfied. I'm gonna put another example to make this more clear.

> Alternative implementation:
> 
> ```
> if get_option('foo_option') # Boolean option
>     foo_dep = dependency('foo_dep')
>     subdir('foo_exec')
> endif
> ```
> 
> Third option: put the dependency() in the subdir.

None of the above solutions would apply if you have multiple subdirs that depend on `foo_dep`. Dependencies should be defined once (in the top meson script) if shared among multiple subdirs. See example in the next comment.

**Comment by adalessandro** [comment]
Here's another example to illustrate a situation with the following requirements:

* Some dependencies are shared among the subdirs, so they should be defined only once, e.g. `dep_a`.
* There's an option for each subdir, that can be enabled independently.
* If a subdir is enabled and can't satisfy its dependencies, should fail and raise an error for that dependency.

```bash
$ tree .
.
├── meson.build
├── meson_options.txt
├── tool_a
│   ├── main.c
│   └── meson.build
└── tool_b
    ├── main.c
    └── meson.build
```

`meson.build`
```meson
project('test', 'c')

# Feature options
dep_a = dependency('dep_a', required : get_option('option_dep_a'))
dep_b = dependency('dep_b', required : get_option('option_dep_b'))
dep_c = dependency('dep_c', required : get_option('option_dep_c'))

if get_option('option_tool_a') # Boolean option
    subdir('tool_a') # Depends on dep_a and dep_b
endif

if get_option('option_tool_b') # Boolean option
    subdir('tool_b') # Depends on dep_a and dep_c
endif
```

So, in this situation I'd force checking for the required dependencies in each subdir script. That way:
* dependency logic is encapsulated for each subdir in its own script.
* dependency objects are defined only once (in the top `meson.build` script), even if shared among different subdirs.
* if the subdir is disabled, no further checking is needed.
* if the subdir is enabled, checking will be performed and fail if not satisfied.

**Comment by eli-schwartz** [comment]
Well, if you check for a dependency twice in a row, it will just reuse the cached dependency. This is an explicitly supported use case. It will result in two log lines, though.

You don't really want to add options specifying the dependencies you want to look for, though -- you want to add options specifying the features you want, and then let them probe for dependencies as needed. This works fine if you check for them inside the subdir, since again -- meson will just cache the dependency() the first time you look for it.

I'd actually suggest being fancy and doing this:

```
project('test', 'c')

# Feature options
require_dep_a = get_option('option_tool_a').enabled() or get_option('option_tool_b').enabled() # .auto() isn't required
dep_a = dependency('dep_a', required : require_dep_a)
dep_b = dependency('dep_b', required : get_option('option_tool_a'))
dep_c = dependency('dep_c', required : get_option('option_tool_b'))

if dep_a.found() and dep_b.found() # dep_b will never be found if tool_a is disabled, only if it is enabled or auto
    subdir('tool_a')
endif

if dep_a.found() and dep_c.found() # dep_c will never be found if tool_b is disabled, only if it is enabled or auto
    subdir('tool_b')
endif
```

Note that you can still use boolean options, and in fact that's what you are doing for dep_a since there's no way to create the union of two feature options. But feature options are *better* if you can fit them into your build description, since they support =disabled as well as =auto, whereas required: get_option('my_bool_option') can only do enabled/auto.

Since in this example each tool has a dependency that takes *full* advantage of the feature option, you can directly rely on the .found() status of that dependency, in combination with dep_a which will always be autodetected even if disabled because it's not using a feature option.

...

tl;dr I still believe feature options are the way to go here.

**Comment by xclaesse** [comment]
There is also: `subdir('foo', if_found: [dep1, dep2, ...])`

**Comment by eli-schwartz** [comment]
True.

But it's still desirable to have `require_dep_a` be a union feature option supporting a disabler.

**Comment by adalessandro** [comment]
Right, but using `subdir('foo', ...)` will still keep the dependency array `foo_deps_array = [dep1, dep2, ...]` defined in the top `meson.build` script (the one calling `subdir()`). Also, that same array `foo_deps_array` will have to be used in the `foo` subdir.

My question is:
* Why having `foo_deps` defined in the top `meson.build` script if that is only specific to `foo`?
* Isn't it more correct to keep `foo_deps` defined in the `foo` subdir only?

**Comment by adalessandro** [comment]
I'm closing the issue as the original point in the RFC is more suitable to be discussed on IRC or mail list. Otherwise, the proposed enhancement doesn't make much sense.
